### PR TITLE
Fix bug in paginable table sorting

### LINKED ISSUE #1861: Sort by columns wrong color and display field name
Please complete the following fields as applicable:

**Expected behaviour:**
'Sort by' is indicated by white ascending/descending arrows

**Actual behaviour:**
'Sort by' is indicated by blue text (e.g. 'Sort by updated_at'). 

**Steps to reproduce:**
- Login
- Look at My Dashboard table(s)

### PR DESCRIPTION
**Pull request by Bodacious**
Fixes #1861 .

Changes proposed in this PR:
- Fix bug preventing paginable links from updating when clicked
- Fix bug: JS was loading multiple nested duplicate .paginable objects
- Refactor Paginable for improved readability.

### CODE REVIEW

File: lib/assets/javascripts/utils/paginable.js [hunk 1]
```diff
@@ -4,14 +4,16 @@ export const paginableSelector = '.paginable';
 
 $(() => {
   const onAjaxSuccessHandler = (e, data) => {
-    $(e.target).closest(paginableSelector).html($(data));
+    $(e.target).closest(paginableSelector).replaceWith($(data));
```
Comment (Bodacious) [inline lib/assets/javascripts/utils/paginable.js]: @briri @xsrust Can you double check this change for me?

The behaviour I saw was that each click on a sort link added a new `.paginable` DOM tree inside the `.paginable` one. I opted for `replaceWith` to fix that, and then moved the event listener to the 'body' element to catch click events on objects just added to the page.

File: app/controllers/concerns/paginable.rb [hunk 2]
```diff
@@ -8,6 +8,9 @@ module Paginable
   # Regex to validate sort_field param is safe
   SORT_COLUMN_FORMAT = /[\w\_]+\.[\w\_]/
 
+  PAGINATION_QUERY_PARAMS = [:page, :sort_field, :sort_direction,
```
Comment (briri) [inline app/controllers/concerns/paginable.rb]: 👍

File: app/controllers/concerns/paginable.rb [hunk 3]
```diff
@@ -101,8 +92,9 @@ def paginable_sort_link(sort_field)
     link_to(
       sort_link_name(sort_field),
       sort_link_url(sort_field),
-      "data-remote": true,
-      class: "paginable-action", "aria-label": "#{sort_field}"
+      class: "paginable-action",
+      data: { remote: true },
+      aria: { label: sort_field }
```
Comment (briri) [inline app/controllers/concerns/paginable.rb]: Didn't realize you could achieve the same result by breaking up the attribute name like this. Will start using this convention going forward 👍

File: app/controllers/concerns/paginable.rb [hunk 4]
```diff
@@ -152,67 +130,62 @@ def refine_query(scope)
...
     end
     <<~HTML.html_safe
       <i class="fa #{class_name}"
          aria-hidden="true"
          style="float: right; font-size: 1.2em;">
 
         <span class="screen-reader-text">
-
-        Sort by #{sort_field.split(".").first}
+          Sort by #{sort_field.split(".").first}
```
Comment (briri) [inline app/controllers/concerns/paginable.rb]: We should probably wrap this for FastGetText `_("Sort by #{sort_field...}")`.

Good catch on the missing `</span>`

File: app/controllers/concerns/paginable.rb [hunk 5]
```diff
@@ -152,67 +130,62 @@ def refine_query(scope)
...
+      query_params[:sort_direction] = sort_direction.opposite
     else
-      sort_url = paginable_base_url_with_query_params(
-        page: page,
-        sort_field: sort_field)
+      query_params[:sort_direction] = sort_direction
     end
-    "#{sort_url}#{stringify_nonpagination_query_params}"
+    base_url = paginable_base_url(query_params[:page])
+    sort_url = URI(base_url)
```
Comment (briri) [inline app/controllers/concerns/paginable.rb]: Good to see the use of URI here

File: lib/assets/javascripts/utils/paginable.js (hunk 1 above)
Comment (briri) [reply lib/assets/javascripts/utils/paginable.js]: If it was nesting them then this should correct the issue and keep the DOM tree clean.
Delegating the listener to the body should work as well. We'll have to test all of the various scenarios to make sure the listeners still work as the table reloads (e.g. sorting after search, the view all link, etc.)

File: lib/assets/javascripts/utils/paginable.js [hunk 6]
```diff
@@ -4,14 +4,16 @@ export const paginableSelector = '.paginable';
...
   };
   $('.paginable-search form[data-remote="true"]').on('ajax:before', e => isValidText($(e.target).find('input[name="search"]').val()));
   // Event listener for Ajax success event captured in response to a paginable link clicked or
   // search form submitted. Note the presence of a selector for on (e.g. a[data-remote="true"])
   // so that descendant elements from .paginable-results that are added in future are also
   // automatically handled.
-  $(paginableSelector).on('ajax:success', 'form.paginable-action[data-remote="true"], a.paginable-action[data-remote="true"]', onAjaxSuccessHandler);
+  $('body').on('ajax:success',
+    'form.paginable-action[data-remote="true"], a.paginable-action[data-remote="true"]',
```
Comment (briri) [inline lib/assets/javascripts/utils/paginable.js]: We should also check pages where there are multiple paginable tables (e.g. My Dashboard) to make sure that firing the success handlers for both tables doesn't cause any issues

File: lib/sort_direction.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+# Examples:
+#
+#   @direction = SortDirection.new(:asc)
+#   @direction.to_s # => "ASC"
+#   @direction.downcase # => 'asc'
+#   @direction.opposite # => 'DESC'
+#
+#   SortDirection.new(:wrong).to_s # => 'ASC'
+#
+class SortDirection
```
Comment (briri) [inline lib/sort_direction.rb]: I really like this. It removes a lot of complexity out of the paginable code.

File: app/controllers/concerns/paginable.rb (hunk 3 above)
Comment (xsrust) [reply app/controllers/concerns/paginable.rb]: neither did I, it's much easier to read though

File: lib/assets/javascripts/utils/paginable.js (hunk 6 above)
Comment (xsrust) [reply lib/assets/javascripts/utils/paginable.js]: Good example of something to check durring UAT. Do we have a process for flagging things like this to @sjDCC and @stephaniesimms to ensure they get checked in UAT?

File: app/controllers/concerns/paginable.rb (hunk 3 above)
Comment (Bodacious) [reply app/controllers/concerns/paginable.rb]: @briri @xsrust  very handy when you have multiple nested attributes such as:

`data: { toggle: "dropdown", target: "#my-dropdown", user_id: 1234 }`

**Pull request merged**